File: src/main/java/gargoyle/ct/cmd/args/CTAnyArgs.java
package gargoyle.ct.cmd.args;

import gargoyle.ct.convert.Converter;

public interface CTAnyArgs {
    <T> T get(Class<T> type, String key);

    <T> T get(Class<T> type, String key, T def);

    <T> T get(int index, T def);

    <T> T get(String key, T def);

    <T> T get(int index, Class<T> type, Converter<T> converter);

    <T> T get(int index, Class<T> type, Converter<T> converter, T def);

    <T> T get(String key, Class<T> type, Converter<T> converter);

    <T> T get(String key, Class<T> type, Converter<T> converter, T def);

    boolean has(int index);

    boolean has(String key);

    int size();
}

File: src/main/kotlin/gargoyle/ct/cmd/args/CTAnyArgs.kt
package gargoyle.ct.cmd.args

import gargoyle.ct.convert.Converter
import kotlin.reflect.KClass

interface CTAnyArgs {
    operator fun <T : Any> get(type: KClass<T>, key: String): T?
    operator fun <T : Any> get(type: KClass<T>, key: String, def: T): T
    operator fun <T : Any> get(index: Int, def: T): T
    operator fun <T : Any> get(key: String, def: T): T
    operator fun <T : Any> get(index: Int, type: KClass<T>, converter: Converter<T>): T?
    operator fun <T : Any> get(index: Int, type: KClass<T>, converter: Converter<T>, def: T): T
    operator fun <T : Any> get(key: String, type: KClass<T>, converter: Converter<T>): T?
    operator fun <T : Any> get(key: String, type: KClass<T>, converter: Converter<T>, def: T): T
    fun has(index: Int): Boolean
    fun has(key: String): Boolean
    fun size(): Int
}
